pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,186 @@
+	_ "embed"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+const (
+	wasmObjectMediaType = "application/vnd.gadget.wasm.program.v1+binary"
+)
+
+type wasmOperator struct{}
+
+func (w *wasmOperator) Name() string {
+	return "wasm"
+}
+
+func (w *wasmOperator) Description() string {
+	return "handles wasm programs"
+}
+
+func (w *wasmOperator) InstantiateImageOperator(
+	gadgetCtx operators.GadgetContext,
+	desc ocispec.Descriptor,
+	paramValues api.ParamValues,
+) (
+	operators.ImageOperatorInstance, error,
+) {
+	return &wasmOperatorInstance{
+		desc:      desc,
+		gadgetCtx: gadgetCtx,
+		logger:    gadgetCtx.Logger(),
+	}, nil
+}
+
+type wasmOperatorInstance struct {
+	desc      ocispec.Descriptor
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	err := i.init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm: %w", err)
+	}
+	err = i.Init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm guest: %w", err)
+	}
+
+	return nil
+}
+
+func (i *wasmOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (i *wasmOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	ctx := gadgetCtx.Context()
+	rtConfig := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(256) // 16MB (64KB per page)
+	i.rt = wazero.NewRuntimeWithConfig(ctx, rtConfig)
+
+	env := i.rt.NewHostModuleBuilder("env")
+
+	if _, err := env.Instantiate(ctx); err != nil {
+		return fmt.Errorf("instantiating host module: %w", err)
+	}
+
+	if _, err := wasi_snapshot_preview1.Instantiate(ctx, i.rt); err != nil {
+		return fmt.Errorf("instantiating WASI: %w", err)
+	}
+
+	reader, err := oci.GetContentFromDescriptor(gadgetCtx.Context(), i.desc)
+	if err != nil {
+		return fmt.Errorf("getting wasm program: %w", err)
+	}
+
+	wasmProgram, err := io.ReadAll(reader)
+	if err != nil {
+		return fmt.Errorf("reading wasm program: %w", err)
+	}
+
+	config := wazero.NewModuleConfig()
+	mod, err := i.rt.InstantiateWithConfig(ctx, wasmProgram, config)
+	if err != nil {
+		return fmt.Errorf("instantiating wasm: %w", err)
+	}
+	i.mod = mod
+
+	// We need to call malloc on the guest to pass strings
+	i.guestMalloc = mod.ExportedFunction("malloc")
```

Comment:
nit: Perhaps, it makes more sense to add this function in `wasm: Expose field functions to guest` commit.

Reply:
I ended up squashing all commits adding the operator. It was so difficult to keep them separated while handling the comments.